Tests/TSCBasicTests/DownloaderTests.swift

```diff
@@ -0,0 +1,252 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import XCTest
+import TSCBasic
+import TSCUtility
+import TSCTestSupport
+import Foundation
+#if canImport(FoundationNetworking)
+import FoundationNetworking
+#endif
+
+class DownloaderTests: XCTestCase {
+    private var downloader: Downloader!
+
+    override func setUp() {
+        let configuration = URLSessionConfiguration.default
+        configuration.protocolClasses = [MockURLProtocol.self]
+        downloader = FoundationDownloader(configuration: configuration)
+    }
+
+    func testSuccess() {
+        let url = URL(string: "https://downloader-tests.com/testBasics.zip")!
+
+        let didStartLoadingExpectation = XCTestExpectation(description: "didStartLoading")
+        let progress50Expectation = XCTestExpectation(description: "progress50")
+        let progress100Expectation = XCTestExpectation(description: "progress100")
+        let successExpectation = XCTestExpectation(description: "success")
+        MockURLProtocol.notifyDidStartLoading(for: url, completion: { didStartLoadingExpectation.fulfill() })
+
+        downloader.downloadFile(at: url, progress: { progress in
+            if progress.spm_isAlmostEqual(to: 0.5) {
```

Comment:
Is this robust?

Reply:
Yes. What are you worried about?

Reply:
I haven't looked closely into the mocking code but I was just wondering how does this work if the progress jumps from 30 to let's say 60.

Reply:
I control the bytes sent